AsyncDisplayKit/ASControlNode.mm

```diff
@@ -354,44 +382,47 @@ - (void)removeTarget:(id)target action:(SEL)action forControlEvents:(ASControlNo
     {
       // Grab the dispatch table for this event (if we have it).
       id<NSCopying> eventKey = _ASControlNodeEventKeyForControlEvent(controlEvent);
-      NSMapTable *eventDispatchTable = _controlEventDispatchTable[eventKey];
-      if (!eventDispatchTable)
+      NSMutableOrderedSet *eventTargetActionArray = _controlEventDispatchTable[eventKey];
+      if (!eventTargetActionArray) {
         return;
-
-      void (^removeActionFromTarget)(id <NSCopying> targetKey, SEL action) = ^
-        (id aTarget, SEL theAction)
-        {
-          // Grab the targetActions for this target.
-          NSMutableArray *targetActions = [eventDispatchTable objectForKey:aTarget];
-
-          // Remove action if we have it.
-          if (theAction)
-            [targetActions removeObject:NSStringFromSelector(theAction)];
-          // Or all actions if not.
-          else
-            [targetActions removeAllObjects];
-
-          // If there are no actions left, remove this target entry.
-          if ([targetActions count] == 0)
-          {
-            [eventDispatchTable removeObjectForKey:aTarget];
-
-            // If there are no targets for this event anymore, remove it.
-            if ([eventDispatchTable count] == 0)
-              [_controlEventDispatchTable removeObjectForKey:eventKey];
+      }
+      
+      NSMutableOrderedSet *newEventTargetActionArray;
+      
+      for (ASControlTargetAction *targetAction in eventTargetActionArray) {
+        BOOL shouldRemove = NO;
+        
+        // Unlike addTarget:, if target is nil here we remove all targets with action.
+        // If target is not nil then check whether it's the same target
+        if (!target || targetAction.target == target) {
+          if (!action) {
+            shouldRemove = YES;
```

Comment:
@Adlai-Holler I ended up using `filterUsingPredicate:`. The result is the same, but I'm happy to rewrite it using your suggestion.